File: src/rpi_4g_streamer/Message.py
"""
Short keys are used here on purpose in order to save space when packing the
messages. Still, the attributes of the classes are descriptive and this is the
only part the devs need to interact with, so this should not be too confusing.
"""

import abc
import time
import msgpack


class Message(abc.ABC):
    """Abstract Base Class for all messages with built-in serialization."""

    _registry = {}

    def __init_subclass__(cls, **kwargs):
        """Automatically register subclasses using their class name."""
        Message._registry[cls.__name__] = cls

    def __init__(self, payload: dict, timestamp: float = None):
        """Initialize message with a dictionary payload."""
        self.timestamp = time.time() if timestamp is None else timestamp
        self.payload = payload

    def to_bytes(self) -> bytes:
        """Serialize the message to bytes using msgpack."""
        return msgpack.packb({
            "t": self.type,
            "p": self.payload,
            "d": self.timestamp,
        })

    @classmethod
    def from_bytes(cls, data: bytes):
        """Dynamically deserialize bytes into the correct message subclass."""
        try:
            obj = msgpack.unpackb(data)
            msg_type = obj["t"]
            timestamp = obj["d"]
            payload = obj["p"]
        except (KeyError, TypeError, msgpack.exceptions.ExtraData) as e:
            raise ValueError("Malformed message payload") from e

        if msg_type in cls._registry:
            return cls._registry[msg_type](**payload, timestamp=timestamp)
        else:
            raise ValueError(f"Unknown message type: {msg_type}")

    @property
    def type(self) -> str:
        return self.__class__.__name__

    def __repr__(self):
        return f"{self.type}(payload={self.payload}, timestamp={self.timestamp})"


class Telemetry(Message):
    """Message type for telemetry data."""

    def __init__(self, v: dict = None, timestamp: float = None):
        if v is None:
            v = {}

        super().__init__({
            "v": v
        }, timestamp)

        self.values = v

    def get_values(self) -> dict:
        return self.values


class Control(Message):
    """Message type for telemetry data."""

    def __init__(self, v: dict = None, timestamp: float = None):
        if v is None:
            v = {}

        super().__init__({
            "v": v
        }, timestamp)

        self.values = v

    def get_values(self) -> dict:
        return self.values


class Syn(Message):
    def __init__(self, timestamp: float = None):
        super().__init__({}, timestamp)


class Ack(Message):
    def __init__(self, timestamp: float = None):
        super().__init__({}, timestamp)


class Latency(Message):
    def __init__(self, timestamp: float = None):
        super().__init__({}, timestamp)


class Command(Message):
    """Message type for command data."""

    def __init__(self, c: str, timestamp: float = None):
        super().__init__({
            "c": c
        }, timestamp)

        self.value = c


class Heartbeat(Message):
    def __init__(self, timestamp: float = None):
        super().__init__({}, timestamp)


"""
Message types used for UDP hole punching
"""


class PeerAnnouncement(Message):
    def __init__(self, r: str, i: str, p: str, timestamp: float = None):
        super().__init__({
            "r": r,
            "i": i,
            "p": p,
        }, timestamp)

        self.role = r
        self.id = i
        self.port_type = p

    def get_role(self) -> str:
        return self.role

    def get_id(self) -> str:
        return self.id

    def get_port_type(self) -> str:
        return self.port_type


class PeerInfo(Message):
    """Message for transmitting peer connection details (IP and ports)."""

    def __init__(self, ip: str, video_port: int, control_port: int, timestamp: float = None):
        super().__init__({
            "ip": ip,
            "video_port": video_port,
            "control_port": control_port,
        }, timestamp)

        self.ip = ip
        self.video_port = video_port
        self.control_port = control_port

    def get_ip(self) -> str:
        return self.ip

    def get_video_port(self) -> int:
        return self.video_port

    def get_control_port(self) -> int:
        return self.control_port

File: android/app/src/main/java/com/v3xctrl/viewer/messages/Message.kt
package com.v3xctrl.viewer.messages

import org.msgpack.core.MessageBufferPacker
import org.msgpack.core.MessagePack
import org.msgpack.core.MessageUnpacker
import org.msgpack.value.ValueType

/**
 * Abstract base class for all messages with built-in serialization.
 * Short keys are used to save space when packing messages.
 */
abstract class Message(
    val timestamp: Double = System.currentTimeMillis() / 1000.0
) {
    /**
     * The message type name, derived from the class name.
     */
    open val type: String
        get() = this::class.simpleName ?: "Unknown"

    /**
     * Pack the payload fields into the MessageBufferPacker.
     * Subclasses must implement this to pack their specific fields.
     */
    protected abstract fun packPayload(packer: MessageBufferPacker)

    @Suppress("UNCHECKED_CAST")
    protected fun packMap(packer: MessageBufferPacker, map: Map<String, Any>) {
        packer.packMapHeader(map.size)
        for ((key, value) in map) {
            packer.packString(key)
            packValue(packer, value)
        }
    }

    private fun packValue(packer: MessageBufferPacker, value: Any) {
        when (value) {
            is String -> packer.packString(value)
            is Int -> packer.packInt(value)
            is Long -> packer.packLong(value)
            is Double -> packer.packDouble(value)
            is Float -> packer.packFloat(value)
            is Boolean -> packer.packBoolean(value)
            is Map<*, *> -> packMap(packer, value as Map<String, Any>)
            else -> packer.packString(value.toString())
        }
    }

    /**
     * Serialize the message to bytes using msgpack.
     * Format: {"t": type, "p": payload, "d": timestamp}
     */
    fun toBytes(): ByteArray {
        val packer = MessagePack.newDefaultBufferPacker()
        packer.packMapHeader(3)

        // type
        packer.packString("t")
        packer.packString(type)

        // payload
        packer.packString("p")
        packPayload(packer)

        // timestamp (seconds since epoch as float, matching Python's time.time())
        packer.packString("d")
        packer.packDouble(timestamp)

        packer.close()
        return packer.toByteArray()
    }

    companion object {
        private val registry: Map<String, (Map<String, Any>, Double) -> Message> = mapOf(
            "Ack" to Ack::fromPayload,
            "Command" to Command::fromPayload,
            "CommandAck" to CommandAck::fromPayload,
            "Control" to Control::fromPayload,
            "Error" to Error::fromPayload,
            "Heartbeat" to Heartbeat::fromPayload,
            "Latency" to Latency::fromPayload,
            "PeerAnnouncement" to PeerAnnouncement::fromPayload,
            "PeerInfo" to PeerInfo::fromPayload,
            "Syn" to Syn::fromPayload,
            "Telemetry" to Telemetry::fromPayload,
        )

        /**
         * Deserialize bytes into the correct message subclass.
         */
        fun fromBytes(data: ByteArray): Message? {
            return try {
                val unpacker = MessagePack.newDefaultUnpacker(data)
                val mapSize = unpacker.unpackMapHeader()

                var messageType: String? = null
                var payload: Map<String, Any>? = null
                var timestamp = 0.0

                repeat(mapSize) {
                    val key = unpacker.unpackString()
                    when (key) {
                        "t" -> messageType = unpacker.unpackString()
                        "d" -> timestamp = unpacker.unpackDouble()
                        "p" -> payload = unpackMap(unpacker)
                    }
                }
                unpacker.close()

                val type = messageType ?: return null
                val factory = registry[type] ?: return null
                factory(payload ?: emptyMap(), timestamp)
            } catch (_: Exception) {
                null
            }
        }

        /**
         * Peek at the message type without full deserialization.
         */
        fun peekType(data: ByteArray): String {
            return try {
                val unpacker = MessagePack.newDefaultUnpacker(data)
                val mapSize = unpacker.unpackMapHeader()

                repeat(mapSize) {
                    val key = unpacker.unpackString()
                    if (key == "t") {
                        return unpacker.unpackString()
                    } else {
                        unpacker.skipValue()
                    }
                }
                unpacker.close()
                "Unknown"
            } catch (_: Exception) {
                "Unknown"
            }
        }

        private fun unpackMap(unpacker: MessageUnpacker): Map<String, Any> {
            val mapSize = unpacker.unpackMapHeader()
            val result = mutableMapOf<String, Any>()

            repeat(mapSize) {
                val key = unpacker.unpackString()
                result[key] = unpackValue(unpacker)
            }

            return result
        }

        private fun unpackValue(unpacker: MessageUnpacker): Any {
            val format = unpacker.nextFormat
            return when (format.valueType) {
                ValueType.STRING -> unpacker.unpackString()
                ValueType.INTEGER -> unpacker.unpackLong()
                ValueType.BOOLEAN -> unpacker.unpackBoolean()
                ValueType.FLOAT -> unpacker.unpackDouble()
                ValueType.MAP -> unpackMap(unpacker)
                ValueType.ARRAY -> unpackArray(unpacker)
                ValueType.NIL -> {
                    unpacker.unpackNil()
                    "null"
                }
                else -> unpacker.unpackValue().toString()
            }
        }

        private fun unpackArray(unpacker: MessageUnpacker): List<Any> {
            val arraySize = unpacker.unpackArrayHeader()
            val result = mutableListOf<Any>()
            repeat(arraySize) {
                result.add(unpackValue(unpacker))
            }

            return result
        }
    }
}
